### PR TITLE
Index.html and animator.html are now separate windows and custom confirm boxes implemented

### PR DESCRIPTION
**Pull request by BoatsAreRockable**
### Custom Confirm Boxes #49
Example of a custom confirm box

As you can see the confirm boxes dim the window and now match the interface! Visually they could be slightly improved - perhaps by making the title a shade of yellow? 

At a later stage a keyboard shortcut should be implemented so that on pressing "enter" the "OK" button is activated. 

Technically I don't see any issues with how I've implemented this but feedback is very much welcome here, so I guess I'll find out.

### Separate Windows #52
This pull request also moves animator.html and index.html into separate windows stopping the menu bar appearing on the latter amongst other things.

And if you are seeing this message then I've successfully learnt how to use branches properly... :smile:

### CODE REVIEW

**Comment by BoatsAreRockable** [comment]
I notice that on Macs and I assume Linux the "OK" and "Cancel" buttons in prompts are usually the other way round to Windows. It would probably best for the order of the 2 buttons to appear different for users of these platforms.

File: app/js/main.js [hunk 1]
```diff
@@ -749,112 +762,157 @@ function notifyError(msg) {
 }
 
 /**
- * Display top menu
+ * Display a custom confirm message
+ *
+ * @param {String|Nunber} [msg=""] The confirm message to display.
```
Comment (le717) [inline app/js/main.js]: Did you install the FuncDocr Brackets extension? It really helps with JSDoc notations.

Comment (le717) [inline app/js/main.js]: Also typo: Nunber

File: app/js/main.js [hunk 2]
```diff
@@ -749,112 +762,157 @@ function notifyError(msg) {
 }
 
 /**
- * Display top menu
+ * Display a custom confirm message
```
Comment (le717) [inline app/js/main.js]: All summaries should end in a period.

File: app/js/main.js [hunk 3]
```diff
@@ -749,112 +762,157 @@ function notifyError(msg) {
...
-if (process.platform === "darwin") {
-    menuBar.createMacBuiltin('Boats Animator');
+function confirmSet(action, args, msg) {
+    "use strict";
+    confirmText.innerHTML = msg;
+    confirmContainer.classList.remove("hidden");
+        
+    // Listen if "OK" is pressed
+    confirmOK.addEventListener("click", function() {
+        window[action](args);
```
Comment (le717) [inline app/js/main.js]: Functions in JS are first-class citizens. You can just pass a reference to a function and call it. Change the code to:

```js
function confirmSet(func, args, msg) {
    // Same code
    confirmOK.addEventListener("click", function() {
        if (args === undefined) {
            func();
        } else {
            func(args);
        }
    });
}
```

Obviously properly indent as needed. It may be possible to do this with a ternary operator but I'm multitasking and can't review this in-depth right now.

File: app/js/main.js [hunk 4]
```diff
@@ -346,7 +359,7 @@ function undoFrame() {
     "use strict";
     // Make sure there is a frame to delete
     if (curFrame > 0) {
-      deleteFrame(curFrame);
+      confirmSet("deleteFrame", curFrame, "Are you sure you want to delete the last frame captured?");
```
Comment (le717) [inline app/js/main.js]: Then with the revised function, this and other calls change to
`confirmSet(deleteFrame, curFrame, "Are you sure you want to delete the last frame captured?");`

The change being the function name is no longer a string.

**Comment by le717** [comment]
I was actually going to extend my revamped notifications code (still WIP, running into issues) to support these, but this works as well. :smile:

I'll let rioforce handle the design, but IMO, it doesn't really match the rest of the design. Something is off (but unsure what).

>I notice that on Macs and I assume Linux the "OK" and "Cancel" buttons in prompts are usually the other way round to Windows. It would probably best for the order of the 2 buttons to appear different for users of these platforms.

Because `process.platform` returns multiple values, we can do this the lazy way for now and expand if we need.

In `startup()`:

```js
if (process.platform === "win32") {
    document.querySelector("body").classList.add("platform-win");
}
```

In your CSS:

```css

/* Or whatever the selectors are */
#btn-ok, #btn-cancel {
  float: right;
}

.platform-win #btn-ok,
.platform-win #btn-cancel {
  float: left;
}
```
 
Basically, we float them right at all times unless we are on Windows, then we float them left for the proper order.

**Comment by le717** [comment]
Wow, you know you are a programmer when you know exactly how to do something and write the code to do it right off the top of your head. :stuck_out_tongue:

**Comment by BoatsAreRockable** [comment]
I forgot about the FuncDocr extension - will install it now (you can probably see that I copied the JSDoc verbatim from one of your notification functions). I also need to install the white space extension you mentioned before. Thank you for your other feedback, you do appear to be some sort of JS wizard!

**Comment by le717** [comment]
You're a wizard, Harry!

**Comment by BoatsAreRockable** [comment]
Implemented @le717's feedback. Don't any obvious reason why this shouldn't now be merged.

**Comment by le717** [comment]
I do. I'd like to do a proper code review for you, and I still think @rioforce should take a look and provide design feedback to make sure everything looks cohesive and matches. :)

**Comment by BoatsAreRockable** [comment]
Okay I shall hold off for a bit then - still getting used to actually being collaborative with this :stuck_out_tongue:

**Comment by rioforce** [comment]
Go ahead and merge it, I'm too busy to style it at the moment, and I can do it after it's merged anyway. It looks solid, though, I haven't tested it yet, I'm sure it's fine.

**Comment by BoatsAreRockable** [comment]
@rioforce I won't await any more design feedback from you before merging then but I will wait for @le717's proper code review first.

**Comment by le717** [comment]
Heads up, code review incoming. I'll say when I'm done. I have to do it tonight because I'll be unreachable until Monday.  Feel free to ask questions, make comments concerns, or whatever when I'm done. I'll get them all and reply back ASAP. :)

File: app/css/animator.css [hunk 5]
```diff
@@ -53,6 +53,47 @@ a {
...
+}
+
+#confirm-title {
+    font-size: 1.2em;
+    color: rgb(224, 224, 100);
+}
+
+#confirm-text { text-align: left; }
+
+#confirm-button-container{ float: right;}
```
Comment (le717) [inline app/css/animator.css]: (I know, I am starting late. You have no idea.) Nit: Missing space before opening brace and after the semicolon.

File: app/animator.html [hunk 6]
```diff
@@ -14,6 +14,18 @@
...
         <p id="msg"></p>
     </section>
+    
+    <!--Confirm messages-->
+    <div id="confirm-container" class="hidden">
+        <section id="confirm">
+            <h2 id="confirm-title"><i class="fa fa-exclamation-triangle"></i> Confirm</h2>
+            <p id="confirm-text"></p>
+            <div id="confirm-button-container">
+                <button id="cancel">Cancel</button>
```
Comment (le717) [inline app/animator.html]: Change ID to `btn-cancel` for consistency with other buttons.

File: app/animator.html [hunk 7]
```diff
@@ -14,6 +14,18 @@
...
+                <button id="OK">OK</button>
```
Comment (le717) [inline app/animator.html]: Change ID to `btn-ok` for consistency with other buttons.

File: app/css/animator.css [hunk 8]
```diff
@@ -53,6 +53,47 @@ a {
...
+}
+
+#confirm-container.hidden { display: none; }
+
+#confirm {
+    position: relative;
+    top: calc(50vh - 10em);
+    display: inline-block;
+    width: 20em;
+    background: #000;
```
Comment (le717) [inline app/css/animator.css]: Should be `background-color`. And to answer your question I never answered, yes, shorthand `background` is used when applied a `background-image` and many of the other `background-*` properties, not when just setting a color.

File: app/css/animator.css [hunk 9]
```diff
@@ -53,6 +53,47 @@ a {
...
+    padding: 0 1em 1em 1em;
+    border: solid 1px #D9D9D9;
```
Comment (le717) [inline app/css/animator.css]: Per spec ([MDN reference](https://developer.mozilla.org/en-US/docs/Web/CSS/border)), `border` shorthand order is `width style color` (e.g., this should read `1px solid #D9D9D9`).

File: app/css/animator.css [hunk 10]
```diff
@@ -53,6 +53,47 @@ a {
...
+
+#confirm-button-container{ float: right;}
+
+#confirm button {
+    float: left;
+    margin-left: 1em;
+    width: 6em;
+}
+
+/*Confirm buttons are typically the other way round in Win32*/
```
Comment (le717) [inline app/css/animator.css]: Nit, not necessary change: One space after opening star and one space before closing star.

File: app/css/animator.css [hunk 11]
```diff
@@ -53,6 +53,47 @@ a {
...
+.platform-win #confirm button { float: right; }
```
Comment (le717) [inline app/css/animator.css]: OK, so I may not have been _exact_ in my code, but I was close enough. :P

File: app/js/main.js [hunk 12]
```diff
@@ -74,18 +74,28 @@ var width  = 640,
 
     // Notification bar
     notifyBar    = document.querySelector(".notification"),
-    notifyBarMsg = document.querySelector(".notification #msg");
+    notifyBarMsg = document.querySelector(".notification #msg"),
+    
+    // Confirm messages
+    confirmContainer = document.querySelector("#confirm-container"),
+    confirmText      = document.querySelector("#confirm-text"),
+    confirmOK        = document.querySelector("#OK"),
```
Comment (le717) [inline app/js/main.js]: Change to `btnConfirmOk` (Same for cancel). Nice job lining up the equals signs. :)

File: app/js/main.js [hunk 13]
```diff
@@ -74,18 +74,28 @@ var width  = 640,
...
+    confirmCancel    = document.querySelector("#cancel");
```
Comment (le717) [inline app/js/main.js]: Since IDs are obviously single-use and be accessed this way, for developer sake to help figure out where in the document this relates to, when possible I like to qualify IDs with the main parent element (even though it is technically unneeded). I'm thinking `#confirm-container #btn-cancel` (and likewise for the OK button).

File: app/js/main.js [hunk 14]
```diff
@@ -128,6 +140,17 @@ function startup() {
...
+    
+    //Load top menu
+    loadMenu();
+    
+    //Maximise window
+    win.maximize();
+    
+    //Windows specific code
+    if (process.platform === "win32") {
+    document.querySelector("body").classList.add("platform-win");
```
Comment (le717) [inline app/js/main.js]: Missing indentation.

File: app/js/main.js [hunk 15]
```diff
@@ -128,6 +140,17 @@ function startup() {
 
     // Set default view
     switchMode("capture");
+    
+    //Load top menu
```
Comment (le717) [inline app/js/main.js]: Comments should have one space after the second slash and before the first character.

File: app/js/main.js [hunk 16]
```diff
@@ -324,10 +347,8 @@ function updateFrameReel(action, id) {
  */
 function deleteFrame(id) {
     "use strict";
-    var confirmDel = confirm("Are you sure you want to delete this frame?");
-
+
```
Comment (le717) [inline app/js/main.js]: Blank line unneeded.

File: app/js/main.js [hunk 17]
```diff
@@ -335,10 +356,7 @@ function deleteFrame(id) {
 
         updateFrameReel("delete", id);
         console.info(`There are now ${curFrame} captured frames`);
```
Comment (le717) [inline app/js/main.js]: With the condition gone, dedent all these statements one level (subject to further review).

File: app/js/main.js [hunk 18]
```diff
@@ -719,7 +737,7 @@ function notifySuccess(msg) {
 /**
  * Display an information notification.
  *
- * @param {String|Nunber} [msg=""] The message to display.
+ * @param {String|Number} [msg=""] The message to display.
```
Comment (le717) [inline app/js/main.js]: I can't believe I typo'd this so badly. -_-

File: app/js/main.js [hunk 19]
```diff
@@ -749,112 +767,163 @@ function notifyError(msg) {
...
+     // Listen if "Cancel" is pressed
+    confirmCancel.addEventListener("click", function() {
+        confirmContainer.classList.add("hidden");
+    });
 }
 
 /**
+ * Display top menu
+ */
+function loadMenu() {
```
Comment (le717) [inline app/js/main.js]: Missing `"use strict;` directive immediately after function declaration. (For more details, I've written [two](https://triangle717.wordpress.com/2015/02/21/js-can-i-stop-useing-strict/) [posts](https://triangle717.wordpress.com/2015/05/06/js-can-i-start-globally-useing-strict/) about it on my blog. /shamelessplug).

File: app/js/main.js [hunk 20]
```diff
@@ -749,112 +767,163 @@ function notifyError(msg) {
...
+    // Create menu
+    var menuBar = new gui.Menu({ type: 'menubar' });
```
Comment (le717) [inline app/js/main.js]: Nit, not important right now: it is all subjective, but the majority of the code uses double quotes, so eventually all quotes should be aligned that way (subject to further discussion if you think single quotes should be used instead ;)).

File: app/js/main.js [hunk 21]
```diff
@@ -749,112 +767,163 @@ function notifyError(msg) {
...
+
+    // Create sub-menus
+    var fileMenuItems = new gui.Menu(),
```
Comment (le717) [inline app/js/main.js]: Eh, I really think `Items` is unneeded on all these.

**Comment by le717** [comment]
Code style review done. Code _functionality_ review is next (I do both types) and I can tell you now, it won't be as long. However, it may be delayed by a short bit. I'll try not to keep you waiting too much.

**Comment by le717** [comment]
Ack, over two hours later! I am so sorry! It may not seem like long to you, but I am pressing to complete stuff because I have a huge day tomorrow and I want to have things done tonight.

I am not totally happy with the current setup of passing the function to be called (you're basically employing a programming setup called callbacks, where a function (or many) is/are passed and the appropriate one(s) is/are called based on the result). I think a feature such as this PR implements in more of the style of `confirm()`. The `confirm()` was intentionally written to return `true` or `false` depending on the button clicked because it is generic and supports many uses cases. While currently our use case is a simple "if yes, run function else do nothing", thinking long term this may not be the case and this code will have to be refactored to support it. I am thinking it would be better to refactor this to copy the generic behavior of `confirm()` to support future use cases.

\*Phew!* That took a while. :sweat_smile: Your very first code review and round 1 is over. Let me know when you complete all that and I'll take another look. And remember, ask questions! You can't learn or understand if you don't ask!

**Comment by BoatsAreRockable** [comment]
I survived round 1! :smiley: Initially I tried to copy the `confirm()` method returning true or false but I struggled to get it to work properly and only return a value after either OK or Cancel was pressed.

You've clearly put a lot of time into reviewing this in detail and I really appreciate it. I will implement the things you've listed ASAP.

**Comment by BoatsAreRockable** [comment]
@le717 I've implemented your style review's feedback. I can understand why a more flexible `confirm()` like solution would be better in the long term. Would the current method be alright for now though?

**Comment by le717** [comment]
I don't know... I'm torn between doing it right the first time or fixing it later when convenient/necessary. I'll try to have an answer by the end of the day (now that my busy weekend is over, I have stuff to do today!).

**Comment by le717** [comment]
Merge it, we'll change it later.
